uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Upgrader.php

```diff
@@ -919,13 +919,86 @@ public function upgrade_1037() {
 
     return TRUE;
   }
-
+  
+  /**
+   * Sets up option values, custom group and custom field
+   * for case type categorization
+   *
+   * @return bool
+   */
+  public function upgrade_1038() {
+    $optionValues = civicrm_api3('OptionValue', 'get', [
+      'option_group_id' => 'cg_extend_objects',
+      'name' => 'civicrm_case_type'
+    ]);
+    if ($optionValues['count'] == 0) {
+      $this->_createOptionValue();
+    }
+    
+    $customGroups = civicrm_api3('CustomGroup', 'get', [
+      'extends' => 'CaseType',
+      'name' => 'case_type_category',
+    ]);
+    if ($customGroups['count'] == 0) {
+      $this->_createCustomGroup();
+    }
+    
+    $customFields = civicrm_api3('CustomField', 'get', [
+      'custom_group_id' => 'case_type_category',
+    ]);
+    if ($customFields['count'] == 0) {
+      $this->_createCustomField();
+    }
+    
+    return TRUE;
+  }
+  
   public function uninstall() {
     CRM_Core_DAO::executeQuery("DELETE FROM `civicrm_navigation` WHERE name IN ('tasksassignments', 'ta_dashboard_tasks', 'ta_dashboard_documents', 'ta_dashboard_calendar', 'ta_dashboard_keydates', 'tasksassignments_administer', 'ta_settings')");
     CRM_Core_BAO_Navigation::resetNavigation();
 
     return TRUE;
   }
+  
+  /**
+   * Creates case type option value for cg_extend_objects option group
+   */
+  private function _createOptionValue() {
+    civicrm_api3('OptionValue', 'create', [
+      'option_group_id' => 'cg_extend_objects',
+      'name' => 'civicrm_case_type',
+      'label' => ts('Case Type'),
+      'value' => 'CaseType',
+    ]);
+  }
+  
+  /**
+   * Creates case type category custom group
+   */
+  private function _createCustomGroup() {
+    civicrm_api3('CustomGroup', 'create', [
+      'title' => ts('Case Type Category'),
+      'extends' => 'CaseType',
+      'name' => 'case_type_category',
+      'table_name' => 'civicrm_value_case_type_category'
+    ]);
+  }
+  
+  /**
+   * Creates category custom field for case type category custom group
+   */
+  private function _createCustomField() {
+    civicrm_api3('CustomField', 'create', [
+      'custom_group_id' => 'case_type_category',
+      'label' => 'Category',
+      'name' => 'category',
+      'data_type' => 'String',
+      'html_type' => 'Select',
+      'is_required' => 1,
+      'column_name' => 'category',
+      'option_values' => array('Workflow' => 'Workflow', 'Vacancy' => 'Vacancy'),
```

Comment:
Instead of hardcoding these options here, would it be possible to just point this to an option group where we can add new option values in the future?

Reply:
I tested again and we can replace the `option_values` with `option_group_id` of an existing record. The ticket requirement was to create these two values. An option could have been to add another implementation for setting up separate option group and values or allow the custom field have setup to auto-generate the group and values at go.